### ISSUE TITLE
Restore brightness of external display which is reset to 100% after restart/resume

### ISSUE THREAD

**Issue by sazconaitbit** [emoacht/Monitorian#115]
Hi,

I'm subscript user from the Microsoft store, I use the app to control my external display brightness is working perfectly, but when I sleep my laptop and wake up the external display brightness is reset to 100% same happen after shutdown.

Would be nice if the app store the latest set value so when i wake up my computer the brightness return to the last value.

Thanks.

**Comment by emoacht** [comment]
Thank you for the suggestion.
Could provide probe.log and operation.log which includes multiple cycles of suspend-resume to get exactly when a reset happens?

**Comment by emoacht** [comment]
Having thought about it, there is a technical difficulty to determine which brightness value is worth to be stored.
First of all, 100% is perfectly valid value and so we cannot tell whether is reset or not from the value itself.
Second, the suspend event by the OS is not reliable. Usually, the OS is too busy when suspend-resume events happen.
Third, the resume event by the OS can be easily filtered out while the succeeding events can not.
Fourth, the manual update by clicking the notification icon can be invoked anytime but there is no clue to tell whether the brightness value is reset at that moment.

**Comment by emoacht** [comment]
I temporarily added the function to restore the brightness which is previously set in this app after reconnect/restart/resume for Microsoft Store version (Ver 2.11). It is accessible from `Restore set brightness` in the menu in the same way as probe.log.

![menu](https://user-images.githubusercontent.com/7205690/105792248-93a48f80-5fca-11eb-8dc0-b048b6d18ec4.png)

I would like to ask any interested user to report the result along with [operation.log](https://github.com/emoacht/Monitorian#operations) which includes the suspend-resume cycles because I am not quite sure what will happens during the cycles in each user's system.

If the function works well, it will be included in add-on features.

**Comment by Joonel** [comment]
I have the same problem but the deal in my monitor. It doesn't save brightness changes so it always resets the brightness after turning off. I think its possible to restore last brightness value after turning monitor on, isn't it?

**Comment by emoacht** [comment]
@Joonel Your idea is basically same as what I implemented.
The technical difficulty is in how to store and determine the last brightness level before PC suspends. For this purpose, it is necessary to store the brightness levels in memory every time when it changes and save them to storage when the PC turns off. Maybe, the consumption of resources for these operations are negligible.
If you try it, please post the logs.

**Comment by Joonel** [comment]
@emoacht 
here is my logs

**Comment by Joonel** [comment]
@emoacht
I turned on 'restore set brightness'. I noticed the function in action sometime, but it seems that it doesn't work properly. 
Also, one of my monitors is losing by Monitorian after its restarting monitor (smth like 'display doesn't support the function'). 

here is my logs [operation.log](https://github.com/emoacht/Monitorian/files/6054481/operation.log)
hope it'll help.

**Comment by emoacht** [comment]
@Joonel Thank you for the log.
Regarding restore function, I could only find successful cases. Could you explain why you think it does not work properly?
Regarding the message, I found that your monitor "G34WQC" has a habit of not responding to the command when 1 minute passed after resume. It's interesting but odd.

**Comment by Joonel** [comment]
Hi.
Monitorian had been working for months somehow expectable, but now I discovered some issues. 
Firstly, it restores the brightness after a PC sleep, but not after login in (a screen is turning off in a few minutes after pressing Win + L so my monitor resets the brightness at next startup). 
Secondly (it may be my special problem), Monitorian is losing my monitor time to time. 
Here is a [log](https://github.com/emoacht/Monitorian/files/6728792/operation.log) file.

**Comment by emoacht** [comment]
Thanks for the log.
Regarding the windows lock, you mean the following part? The brightness changed 4 times in row in a very short period.
```
[Date: 27.06.2021 17:30:44 Ver: 2.19.0.0]
HandleBrightnessChange "G34WQC"
 BEF [Save] 65 - 65
 AFT [Save] 65 - 65

[Date: 27.06.2021 17:30:47 Ver: 2.19.0.0]
HandleBrightnessChange "G34WQC"
 BEF [Save] 65 - 95
 AFT [Save] 95 - 95
HandleBrightnessChange "G34WQC"
 BEF [Save] 95 - 100
 AFT [Save] 100 - 100
HandleBrightnessChange "G34WQC"
 BEF [Save] 100 - 70
 AFT [Save] 70 - 70
HandleBrightnessChange "G34WQC"
 BEF [Save] 70 - 100
 AFT [Save] 100 - 100

[Date: 27.06.2021 17:44:43 Ver: 2.19.0.0]
CheckTransition (IsTransition: True)

[Date: 27.06.2021 17:44:43 Ver: 2.19.0.0]
OnMonitorsChangeInferred (PowerWatcher- Suspend 0)
```

Regarding the detection, the monitor seems not to be controllable sometimes. It cannot be helped by this app.

**Comment by emoacht** [comment]
I am not sure when those changes happened but going to change not to save the brightness after the OS locks and to restore it when the OS unlocks.

**Comment by Joonel** [comment]
@emoacht Now it works as expected (brightness is restoring after PC unlock 🥳) except times when my monitor isn't responding... I didn't think that a cheap monitor could be bad in that simple function.

**Comment by emoacht** [comment]
This restore function is improved to take into account monitors' unison state. See #271

**Comment by githubyouser** [comment]
This is still an issue for me, even with "restore brightness on reconnection" checked. I don't remember having this issue with Windows 10 and the same monitor, but after getting a new laptop with Win 11, my external monitor always comes on *way* too bright after resuming from sleep, and I have to manually adjust it every single time back to where it is supposed to be. Logs attached. Let me know if there's any other troubleshooting I can help with. 
[operation.log](https://github.com/emoacht/Monitorian/files/10197090/operation.log)
[probe.log](https://github.com/emoacht/Monitorian/files/10197091/probe.log)

**Comment by emoacht** [comment]
Thanks for the log.
But the operation.log is too short to track changes. Could you repost the longer which includes multiple disconnect and reconnect of "Q24i-10"?

**Comment by githubyouser** [comment]
Here's a longer log. [operation.log](https://github.com/emoacht/Monitorian/files/10201173/operation.log)
I often have to "bump" the brightness of "Q24i-10" quickly up and then back down to make it drop back to where it's supposed to be. (I keep it at 0 normally, and when the laptop resumes, it *shows* 0 but is clearly brighter. So I just bump it up a little and then back down, and it goes to the proper brightness.)

**Comment by emoacht** [comment]
`it shows 0 but is clearly brighter`
Wierd. If a monitor does not report correct value, it will be difficult to handle such monitor.

**Comment by githubyouser** [comment]
Yeah, it is kinda weird, but what's strange is that I don't remember this ever happening with my old laptop (running Windows 10) with the same monitor. So I'm not sure if it's a Windows 11 bug, or if it's some sort of driver issue that didn't exist on my older laptop.

**Comment by emoacht** [comment]
I added an experimental code to Microsoft Store version Ver 3.15.1. This version will ignore the brightness reported by a monitor when restoring the brightness.

So, could you udapte to this version and check if this code makes any difference? To enable it, you need to input `/restore hard` in [Command-line arguments](https://github.com/emoacht/Monitorian#command-line-arguments).

**Comment by githubyouser** [comment]
Awesome, thanks! After inputting the command line arguments and restarting Monitorian the brightness goes down to 0 without having to manually "bump" it up and back down. It also seems to be working when I wake up the computer from sleeping. 

However, it still doesn't seem to work when the computer **_screen_** times out (after 5 minutes). The computer isn't actually sleeping, it just shuts off the display. When I wiggle the mouse to turn the display back on, then the external monitor wakes up at high brightness and does not drop down to 0.

**Comment by emoacht** [comment]
> However, it still doesn't seem to work when the computer screen times out (after 5 minutes). The computer isn't actually sleeping, it just shuts off the display.

This restore function is not originally intended for these events but there no particular problem in covering it.
Could you try Ver 3.15.2 in the same manner?

**Comment by githubyouser** [comment]
I think it's working! 🥳 The external monitor "wakes up" too bright, but after a few seconds it dims back down to 0. 

Thank you so much! I don't think I've ever had such a quick response/fix to a software issue before. I really appreciate your time and skill to help me out. 😄

**Comment by banana99ou** [comment]
Hey, can you make contrast stay where i set it after reboot?

**Comment by MrChriZ** [comment]
I'm also seeing this - Windows 11, V4.10.2, iiyama Prolite screen.
Mine always wakes up dim!  (bit like me really!)

**Comment by kshitij-garg** [comment]
Hi everyone! 👋 Just wanted to drop a quick note that if you're looking for an immediate fix/implementation for this feature, I've compiled a suite of these quality-of-life updates into an unofficial fork called [Monitorian 2.0 Unofficial](https://github.com/kshitij-garg/Monitorian-2.0).

You can easily download the ready-to-run .exe from the [Latest Release Page](https://github.com/kshitij-garg/Monitorian-2.0/releases/tag/v2.0.0).

I've submitted this feature as a PR upstream, but in the meantime, feel free to use the unofficial build if you need it today. Huge thanks to @emoacht for the amazing original software!

**Comment by emoacht** [comment]
@MrChriZ The operation.log you posted is too short and does not include information showing the reported issue.
If your problem still exist, please post the longer one?
